addons/account/models/account_move.py

```diff
@@ -2561,20 +2561,51 @@ def _is_readonly(self):
     # -------------------------------------------------------------------------
     def _is_eligible_for_early_payment_discount(self, currency, reference_date):
         self.ensure_one()
-        payment_terms = self.line_ids.filtered(lambda line: line.display_type == 'payment_term')
-        return self.currency_id == currency \
-            and self.move_type in self._early_payment_discount_move_types() \
-            and self.invoice_payment_term_id.early_discount \
-            and (
-                not reference_date
-                or not self.invoice_date
-                or (
-                    (existing_discount_date := fields.first(payment_terms).discount_date)
-                    and
-                    reference_date <= existing_discount_date
-                )
-            ) \
-            and not (payment_terms.sudo().matched_debit_ids + payment_terms.sudo().matched_credit_ids)
+        return self._is_eligible_for_early_payment_discount_batched(currency, reference_date)[self.id]
+
+    def _is_eligible_for_early_payment_discount_batched(self, currency, reference_date):
+        is_eligible_by_move_id = {move.id: False for move in self}
+        move_types_for_epd = self._early_payment_discount_move_types()
+        eligible_moves = self.filtered(lambda m:
+                                       m.currency_id == currency
+                                       and m.move_type in move_types_for_epd
+                                       and m.invoice_payment_term_id.early_discount
+                                       )
+        if not eligible_moves:
+            return is_eligible_by_move_id
+
+        payment_term_lines = self.env['account.move.line'].search([
+            ('move_id', 'in', eligible_moves.ids),
+            ('display_type', '=', 'payment_term'),
+        ])
+
+        reconciled_pt_lines = payment_term_lines.filtered(
+            lambda line: line.sudo().matched_debit_ids or line.sudo().matched_credit_ids
+        )
+        reconciled_move_ids = set(reconciled_pt_lines.move_id.ids)
+
+        pt_lines_by_move = {}
+        for line in payment_term_lines:
+            if line.move_id.id not in pt_lines_by_move:
+                pt_lines_by_move[line.move_id.id] = self.env['account.move.line']
+            pt_lines_by_move[line.move_id.id] |= line
```

Comment:
Weird to do this in a perf PR when this is O(n²) 😆 
Doesn't `grouped` exist in 18.0?